### DISCUSSION TITLE
Astro middleware

### DISCUSSION THREAD

**Discussion by pilcrowonpaper** [withastro/roadmap#469] [Proposal]
# Summary

Introduce a middleware to Astro, where you can define code that runs on every request. This API should work regardless of the rendering mode (SSG or SSR) or adapter used. Also introduces a simple way to share request-specific data between proposed middleware, API routes, and `.astro` routes.

# Background & Motivation

Middleware has been one of the most heavily requested feature in Astro. It's useful for handling common tasks like auth guards and setting cache headers. For me, it would make handling authentication much easier.

This proposal is heavily inspired by [SvelteKit's `handle` hooks](https://kit.svelte.dev/docs/hooks#server-hooks-handle).

# Goals

- Provide a way intercept requests and responses, allowing users to set cookies and headers
- Works both in SSR and SSG mode
- Allow users to use community-created middlewares (libraries)
- Provide an API for request-specific data

# Non-goals

- Route specific middleware

# Example

[A quick prototype](https://www.typescriptlang.org/play?target=99&jsx=0#code/PTAED1QJQVwO1AFwBYFNQGMD2ATVBCAKEMQE8AHdKVAZywBsA3dAXlAAps5FUAPRAFzRUARxi1EAYSzc+iAJSgWAPlAAFAE5YAtgEsaqADzUa5GQeUkK6ALK6cOeqgDuAQw2sOXHvyHUxEtKy-AA0oB50TKh+tAzMiirqWnoGxrRmcBZWlMIBNFIyPohKoADehKCV4aLi+TF5iCEVVdhYANa6tEI2ruSG+Rq6cADmYQNDw8pNVaD0WBiu9DQx2Bo4hq5wpGGbpJYzIKDk9DA0oFgoqBqg+TAAZneEAL7EXPmgeg5ObhHd9o4udyoADaAF0SsDmpVXDRSHAMF5CnIwhE4qgEqpyjMZm8GKgAHRzYbsABEAEZXCT5FDsd45IT5osaPjTlc4K5tJ4ScgsKdEDIAPoAIwuJJpOPMxQiGQMJVcbl0UtiUU4SP41OxLXMeIZxPJQqp4qqHkQMA0CGl5lQNKe0yqMLhCNVwUa1Ui8SUmKNlVxTl1pIATJSNZrMJK3TLPPLXIq3WjnUUQ5rfQSiYGDUnsSazRb0labXbobD4YiXSjlR7EljQyn-SSAMzB71hzJK0xWuUKtvxunq5u1tMNjPN7PmiP5ma2wig4iHClhIMLoVhMlC17h1FRKC8nhy4tOjwNeq1BSeso0t7dqIAQQGWG3MF3bAdJYTcmPgTVp6rI9QprHcAuMI7aZKg7AAAaGMgZLKGg9BzKAAAkpS9ogDILEsLIGOaHKoE8hjANB4FhNWoZVPkrimssoABgADLRhZkaAaCuHgGjUaRTEzCSQQ8NwAC0AAq1gkkIJJFIRiDaPQADcmDIO4BiICwj53PxAAcYpcZULxkU8ma6VqrYtkUH75LxcglJxMyHiejESu0nTUYBzigD05DsPI9lVHMGEcYZOkXuGwx-iYkYlJ8AI-LQ+IeMw7FgfIsWoDgMAYGB7CbswMTuqgYTaM4GLnmRo4IC+Tq+YsFnqmeBXsJV9DVa6WXogWcY3neD48Jml7jqBnYxsUIWIGFVpvn2EqZDqabgR4cBsSlzFSfQQjIdGsaWqB+JFJ5TzgdSLyEL1QzYHoIz+CeJQubkJ7sNdACqUAADKksgiCIOQAggA1PJ1PW9G0VS1ItV1YEnToEwXRI8hAA)

Middlewares can be defined in `src/middleware.ts` by exporting `middleware` (array):

```ts
export cost middleware = []
```

A simple middleware looks like so:

```ts
export const middleware: Middleware = async (context: APIContext, resolve: Resolve) => {
  const session = await getSession(context.request);
  const isProtectedRoute = matchRoute(context.url);
  if (!session && isProtectedRoute) {
    // early response
    return new Response(null, {
      status: 400,
    });
  }
  context.cookies.set("random-cookie", "some-value");
  const response = await resolve(context); // resolve api route or render page
  return response;
};
```

`context ` is the same one provided to API route handlers. Most of Astro's request handling process will be behind `resolve()`, and the response object returned from `middleware` will be sent to the user's browser. 

## Multiple middleware

`sequence` can be imported to run multiple middlewares in sequence.

```ts
export const middleware: Middleware = sequence(
  async (context, resolve) => {
    console.log("1a");
    const response = await resolve(event);
    console.log("1b");
    return response;
  },
  async (context, resolve) => {
    console.log("2a");
    const response = await resolve(event);
    console.log("2b");
    return response;
  }
);
```

The log result for this example is:

```
1a
2a
2b
1b
```

## `locals`

This proposal also adds `locals` property to `APIContext` and `AsroGlobal`. This `locals` object will be forwarded across the request handling process, allowing for data to be shared between middlewares, API routes, and `.astro` pages. This is useful for storing request specific data, such as user data, across the rendering step.

```ts
export const middleware = async (context: APIContext, resolve: Resolve) => {
  context.session = await getSession(context.request);
  const response = await resolve(context);
  return response;
};
```
```astro
---
// pages/index.astro
const session = Astro.locals.session;
---
```
The value type of `locals` can be anything as it won't be JSON-stringified:

```astro
---
// pages/index.astro
const session = await Astro.locals.getSession();
---
```

`locals` can be typed inside `src/env.d.ts` (I think it's possible?):

```ts
// src/env.d.ts
// at least one of these should be possible
type Locals {}

declare namespace App {
    type Locals = {}
}
```

## SSG mode

The middleware function will run on each route's pre-rendering step, as if it were handling a normal SSR request.

```ts
export const middleware = async (context: APIContext, resolve: Resolve) => {
  const response = await resolve(context); // pre-render step here
  return response;
};
```

**Comment by pilcrowonpaper** [comment]
One common pattern for middleware found in Express and Next.js looks something like this:

```ts
const middleware = (req, res, next) => {
  // do stuff
  if (condition) {
    next();
  }
};
```

However, this is only possible because Node native request/response API is used. Since Astro uses web standard `Request` and `Response`, a different approach was needed.

**Comment by natemoo-re** [comment]
I like this a lot! Thanks for sharing the proposal. 🙌🏻 

Enforcing namespaces is something I'd want to factor into the proposal to avoid collisions as much as possible. Type-safety is also very important for DX, so I'd love if we could do this in a type-safe way.

Here's a quick experiment in that vein on the [TypeScript Playground](https://www.typescriptlang.org/play?target=99&jsx=0#code/PTAED1QJwVwO1AQzgE1AYwBYFN0GtQAXHUAGwHsBzAZwCgVdTEptQBbclGU1gckWqEo5AFxsAlihQ8A7s2y9QAb1qhQ2AB4AHclEKhxcQtigAzROlYAlbJXGCTygL6rQryhQBGiUstdrCAE8tVgBBQWEAWUlpbDkWABFEQkRQAF5QGzsHKHUNY1RqTNxdFAAeQ1NHABUAOUQ2bC1qC2wAGgM4KtzqpJS8gpQim3RSsoBpbECO5ECAPjnQAH5QesbqLVaUSMQtMrqGppbLDt7kxEWRVcONrZ29yenQOGwANxM51xcXWiCQ0Ee6VAEUMlAAPnAYGxPCYwdRAtDyKQANy0dBMahFNbYW6Wba7fbY5qtAbYQoAqanPqpTSDYYlKDlR4zODzRYqNRqADEKHOVxGYwOjWJJ1AZxSiwySicSCKs1RnOB2EIAAo4DdNpYrkKjq0Oq8fDBsNrqQBKK6vciSBWc6jK-aA2lkoagPBTcimMXUuZqjWtbVE47tV1TbXM0AG0hGk3nADa1UeAF1zRGrSgbWo7ar1etNcaxYG9aAAHSl5g0K6zFOWyR+RWK8SelXl6jFnhwSjEdJpDIABlNREwwhkzzioAAolBhFAVbxqgB5BLz3imjP1xugZtQGhtsmdzDdjIARgHHPr59GcEEoFjkaNiaBLbX54CmHsxZ5cZzOLzD4yd+wZ8XxYQgYCgOAgLUFwXwMJsW13Dsux7DIACZT38GDQBAMVF3nK5MGQWJQBQch9HhIxEA0DCYMva9YzdJ4AL-JBtzlIp6PdT1xUQKk4wYj0vXORNE0g+tiHfT8Uljb9cWwRMOMCZiANExUQLAiDqM5aCtNoDDKGVX1c39AtDhFbAU24td9MIB0plJcl+K471DJ-YydTMjoGLDKYLOpeMkysgyZLzANTKDDpS2LFtK1ZFNZjrF8Ny3Hd233Q9QH7aBlXUwcJN5KTgtaETNLUbCFyXfDCJ4YjSOBQIKKomjyCvfQFOYltZRvRzBJSYqYLU8DctbSTEGkv1LHkhi+p0n4-msHEkXeIEVUvYx8n5bAAEcjUEABhZq1sIAc0kWAAFYQJDtMobA2Zq7U+ObQGiKRZHkZbVtpDbtpxQh9qMWkOhYahFvzG6QeOs6LvsbBrp-O7sAe4J5u+vaDtpIEzzUFgUcIL6dsINoMNGcg8HEHErnuMoQQ7DpqcoOZCcVfLECucIhHIZ7YnibBqQw7CtCjIpSJwXJBBgUxTFoH5aP0CQXrieRqApmJXpYWNmNjDCBHq9BNw+-JAYW0h3ghhKLzR-Ji2Z4ss1nAABAR2YAK2oYBEBgYheA6JRQBgO1wMOK5eEwcg-cIZqAH1PFIr2kA9nAjHEdBkmwFAriEI1QCcVcSowO79CBnQr1YDJEDkcQC6N94Voto7c9opFsDbKgVQAIiPTxW9NXOBoQQv4YwpxGc5bW4F1mv-oNrLgeN8z0nZeu7sb5vKDblDEC7xeWunou7UfcvK5n6v9brpqr2XihV9blDO+7-rssG-vi8H4e1FH8eT8No+55Os36wbngK824AGYN53xfDLHe8N96IArtPEGE9DrgPNufQBl8QG3x7g-PucNn6KiHrQRMulsJHh4qAdeHQb4dA7rpSBQMQZWFDsYR85Fx7Y3xnjH6ptMZ523vQ2ebNhCMI9iXWUOs9a104ajSeR155-0VJAn27tiBkkIEnFOKAOh+xMN+LOQIT5W3OMWay9tHbCBdm7eOK5RJJQAITKITmo5OxgUDoUwpyXuo4Rw3V3tgFUAApAAyvOWoNshCgkbIEFUPsTDTiuK3WotUHGqPUS41uWdTTe1zjBQQyQ-ZXAACy9mAa-dxnIcCIAYFAJW8iymKlbn9AohAAC01Qkat3ibsAWqTxDNWAC7ZqyIMAEWqcqNIHtTDNIAByt2yS+bS7js65wWfWTxLxvG4LtCqAABmUTAR45g4FIBQUAAASJQ2iA6NCcGUYA+y5jbKyXU3JoEakoV7L2UpMEKlVJqTwupDS0ZGFae0+Jh07mEDYCiYZzAszjMIJMmZcyoLLOQVnIm+deGHSkb9WuGMsE4y+ZyYmpNyZeKersFUmTc7MyuOs64Rk8T3CpYPDF29rI+OgRkOWXNFbFhYO8UZVL+WpxgJYFUKp+HvA2t-DobAZDcKwaBQa79NwUGTqQRptJTbypVOqnwWqp5SvMi-eBAiIjkGEcYNFdDNmiLLrA-QHK7WIO1WymeTd0HbKfnaK45yHVwJ9U3Q6VKnDbO7tLTFhhRgSA7DYHGQJ6XxvxmqMcABVKwAAZNumBCCEC0CIEA+rSAh0ECIYBHzexd27saq1fjo3kFjZQZNXDdIMHRG9DgXBqr8AtWIFWCsWCKDPIYYwZgSRZHsGOv+rcHYWosQ4jptTLnfiuHTICyTE7ONTlcaOjdkBrm0t8IAA). I've implemented "locals" (renamed to "data") with namespacing.

Providers (middleware) could register their "data" using the following pattern, ensuring that their data is properly namespaced.

```ts
interface AuthData {
    user: { name: string };
    authenticated: boolean;
}

declare module 'astro:middleware' {
  interface Register {
    "@acme/auth-middleware": AuthData
  }
}
```

Consumers (user, other middleware) could then read from "data" in a type-safe way.

```ts
// Get all data
const { authenticated, username } = context.data.get('@acme/auth-middleware');
// Get specific nested key
const username = context.data.get('@acme/auth-middleware', 'user.name');
```

This is possibly by using a `NamespacedMap` class, which could look something like:
```ts
type Key = string|number|symbol;
class NamespacedMap<TNamepsace extends Key, TData extends Record<Key, any>> {
    #data: Record<TNamepsace, TData> = {} as any;
    set(namespace: TNamepsace, value: TData): void;
    set<TKey extends keyof TData>(namespace: TNamepsace, key: TKey, value: TData[TKey]): void;
    set(namespace: TNamepsace, ...args: any): void {
        if (args.length === 0) throw new Error('TODO');
        if (args.length === 1) {
            const [value] = args;
            this.#data[namespace] = value;
            return;
        }
        if (args.length === 2) {
            // TODO: handle dot syntax
            const [key, value] = args as [keyof TData, TData[keyof TData]];
            this.#data[namespace][key] = value;
            return;
        }
    }

    get(namespace: TNamepsace): TData;
    get<TKey extends keyof TData>(namespace: TNamepsace, key: TKey): TData[TKey];
    get(namespace: TNamepsace, ...args: any): any {
        if (args.length === 0) return this.#data[namespace];
        // TODO: handle dot syntax
        const [key] = args as [keyof TData];
        return this.#data[namespace][key];
    }
}
```

**Reply by pilcrowonpaper** [reply]
I actually like the idea of using maps! It's like an import statement, but for using data from middleware instead of modules.

I think `NamespacedMap.set` doesn't have to be complicated though, and it should be limited to just a key value pair. There's no way for TS to check that the middleware set every properties registered if you allow values to be defined using dot syntax:

```ts
context.data.set('@astrojs/auth', "username", "houston_bot");
// ts doesn't realize I didn't set "@astrojs/auth.authenticated"
``` 
```ts
context.data.set('@astrojs/auth', {
  username: "houston_bot" // ts error - you're missing "authenticated"!
})
```

As for the name, I don't have strong preferences but I chose `locals` since people who've used Express or SvelteKit will feel familiar, and it's a name that is easy to remember but still distinct.

**Reply by pilcrowonpaper** [reply]
That said, I think `locals`/`data` can also be used to pass data between `.astro` components as well, so I don't think it should be middleware specific.

**Reply by matthewp** [reply]
I think this makes the API significantly more cumbersome and collisions aren't super likely to begin with. I think this is a problem to solve once we encounter it, not one to anticipate.

We should try to figure out a way to make `Astro.locals` strongly typed, if possible though. Maybe a middleware provider can modify the type to add their own data properties?

**Comment by thdxr** [comment]
This is something we would require to allow for some of the magic that SST does to make sure things work the same in dev and when released.  We have a whole hooks system that needs to be booted up when an request comes in

**Comment by jamesarosen** [comment]
I built and maintained a middleware system for use in Cloudflare workers, which uses the service-worker environment. The API for that system was `(FetchEvent, (FetchEvent) => Response) => Promise<Response>`. Some examples:

```ts
export async function preventFraming(fetchEvent: FetchEvent, next: Middleware) {
  const upstreamResponse = await next(fetchEvent)
  if (!upstreamResponse.headers.has('x-frame-options') {
    upstreamResponse.headers.set('x-frame-options', 'DENY')
  }
  return upstreamResponse
}

export async function adminAuth(fetchEvent: FetchEvent, next: Middleware) {
  const { pathname } = new URL(fetchEvent.request.url)
  if (!/^\/admin\//.test(pathname)) return next(fetchEvent)

  const rawJWT = fetchEvent.request.headers.get('jwt')
  if (!rawJWT) {
    return new Response('Unauthorized', { status: 401 })
  }

  const jwt = await parseJWT(rawJWT)
  if (jwt?.payload?.roles?.indexOf('admin:read') ?? -1 === -1) {
    return new Response('Forbidden', { status: 403 })
  }

  return next(fetchEvent)
}
```

I like this API because it's closely tied to open standards (`Request`, `Response`, `FetchEvent`), small, and very flexible. We can add extra context to `fetchEvent` if we want to expose Astro APIs to the middleware.

They also compose easily with `reduce`:

```
const BOTTOM_OF_STACK = ({ request }): FetchEvent => fetch(request) // or Astro.handle or whatever

function compose(...stack) {
  return stack.reduce((next, current) => (e) => current(e, next), BOTTOM_OF_STACK)
}
```

**Comment by matthewp** [comment]
I like this idea a lot, I think it's on the right track. Couple of questions:

1. How can adapters provide data for `locals`? One use-case we hear about a lot is wanting runtimes like Cloudflare which have their own special APIs to add data that components can then use.
2. We should talk about the different ways to register a middleware. The `src/middlewares` is what you suggest here, but I could also see this being done via integrations and/or config. `middlewares: []`. Maybe there are other ways.

**Reply by matthewp** [reply]
edit: scratched out a question that is covered in the non-goals section

**Comment by matthewp** [comment]
This proposal has been accepted and has moved into stage 2: https://github.com/withastro/roadmap/issues/531
